staging/vhost-device-can/src/can.rs

```diff
@@ -0,0 +1,383 @@
+                        // Regular CAN frame
+                        trace!("Received CAN message: {:?}", frame);
+                    }
+                    CanAnyFrame::Fd(frame) => {
+                        // CAN FD frame
+                        trace!("Received CAN FD message: {:?}", frame);
+
+                        let read_can_frame = VirtioCanFrame {
+                            msg_type: VIRTIO_CAN_RX.into(),
+                            can_id: frame.raw_id().into(),
+                            length: (frame.data().len() as u16).into(),
+                            reserved: 0.into(),
+                            flags: frame.id_flags().bits().into(),
+                            sdu: {
+                                let mut sdu_data: [u8; 64] = [0; 64];
+                                sdu_data[..frame.data().len()].copy_from_slice(frame.data());
+                                sdu_data
+                            },
+                        };
+
+                        match controller.push(read_can_frame) {
+                            Ok(_) => warn!("New Can frame was received"),
+                            Err(_) => {
+                                warn!("Error read/push CAN frame");
+                                return Err(Error::SocketRead);
+                            }
+                        }
+                    }
+                    CanAnyFrame::Remote(frame) => {
+                        // Remote CAN frame
+                        trace!("Received Remote CAN message: {:?}", frame);
+                    }
+                    CanAnyFrame::Error(frame) => {
+                        // Error frame
+                        trace!("Received Error frame: {:?}", frame);
+                    }
+                }
+
+                controller
+                    .rx_event_fd
+                    .write(1)
+                    .expect("Fail to write on rx_event_fd");
+            }
+        }
+    }
+
+    pub(crate) fn exit_read_thread(&mut self) {
+        trace!("Exit can read thread\n");
+        self.status = false;
+    }
+
+    pub(crate) fn config(&self) -> &VirtioCanConfig {
+        trace!("Get config\n");
+        &self.config
+    }
+
+    pub(crate) fn can_out(&self, tx_request: VirtioCanFrame) -> Result<u8> {
+        trace!("Can out\n");
+
+        // Create a CAN frame with a specific CAN-ID and the data buffer
+        let can_id = StandardId::new(tx_request.can_id.to_native().try_into().unwrap())
+            .expect("Fail to create StandardId");
+        let data_len = tx_request.length.to_native() as usize;
+
+        let data: Vec<u8> = tx_request.sdu.iter().cloned().take(data_len).collect();
+        let frame = CanFdFrame::new(can_id, &data).expect("Fail to create CanFdFrame");
+
+        // Send the CAN frame
+        let socket = self.can_out_socket.as_ref().ok_or("No available device");
+
+        match socket {
+            Ok(socket) => match socket.write_frame(&frame) {
```

Comment:
I think `write_frame()` shall not be a blocking call because will block the daemon but I am not sure if that may be an issue. Sometimes if the daemon is blocked, it can't respond to control cmds and that may trigger a timeout in the guest. I am not sure if this may happen in in the can driver.

Reply:
Indeed, ideally we could eliminate that case. As far as I know, virtio-can doesn't have any timeout restrictions into the kernel, and control messages are not exchanged so oftenly as in virtio-sound case which uses them for starting and stopping each audio stream. In CAN case, ctrl messages are spotted during the setting up/down process of the guest interface. Based on that I think, if you agree, we can leave it as it is for the moment, and revisit it in the future. Let me know your thoughts on this

Reply:
I think we can leave like this. I just wanted to highlight that may be not a good idea to call functions that could block. I do not know if this is the case for `write_frame()`, maybe, it is not.

Reply:
Hello Matias, going through the spec pointed me the "VIRTIO_CAN_F_LATE_TX_ACK" feature, which is used to make the device mark the new tx_request as used after complete writing to the host's interface. In the current implementation that's always the case. Ideally we would like in the future, when this feature is disabled, to make that write asychnonously and return immediately.